flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceScanTestPrograms.java

```diff
@@ -0,0 +1,259 @@
+                                    .build())
+                    .runSql("INSERT INTO sink_t SELECT a, b, c FROM source_t LIMIT 2")
+                    .build();
+
+    static final TableTestProgram PARTITION_PUSHDOWN =
+            TableTestProgram.of(
+                            "table-source-scan-partition-pushdown",
+                            "validates table source scan with partition pushdown")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("source_t")
+                                    .addSchema("a INT", "b BIGINT", "c VARCHAR")
+                                    .addPartitionKeys("b")
+                                    .addOption("partition-list", "b:1,b:2")
+                                    .producedBeforeRestore(BEFORE_DATA)
+                                    .producedAfterRestore(AFTER_DATA)
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink_t")
+                                    .addSchema("a INT", "b BIGINT", "c VARCHAR")
+                                    .consumedBeforeRestore(
+                                            "+I[2, 2, hello]", "+I[3, 2, hello world]")
+                                    .consumedAfterRestore("+I[5, 2, foo bar]")
+                                    .build())
+                    .runSql("INSERT INTO sink_t SELECT a, b, c FROM source_t WHERE b = 2")
+                    .build();
+
+    static final TableTestProgram READING_METADATA =
+            TableTestProgram.of(
+                            "table-source-scan-reading-metadata",
+                            "validates table source scan by reading metadata")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("source_t")
+                                    .addSchema(
+                                            "a INT",
+                                            "b BIGINT",
+                                            "c VARCHAR",
+                                            "d TIMESTAMP(3) METADATA")
+                                    .addOption("readable-metadata", "d:TIMESTAMP(3)")
+                                    .producedBeforeRestore(BEFORE_DATA)
+                                    .producedAfterRestore(AFTER_DATA)
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink_t")
+                                    .addSchema("a INT", "c VARCHAR")
+                                    .consumedBeforeRestore(
+                                            "+I[1, hi]", "+I[2, hello]", "+I[3, hello world]")
+                                    .consumedAfterRestore("+I[4, foo]", "+I[5, foo bar]")
+                                    .build())
+                    .runSql("INSERT INTO sink_t SELECT a, c FROM source_t")
```

Comment:
This will not read the metadata, will it?

Reply:
Thanks. Fixed it by selecting the metadata column in the query